<a href="https://colab.research.google.com/github/medulka/LLMs/blob/main/gemma_2B_finetuning_LoRa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
print("hello world!")

hello world!


In [ ]:
# 15.05.2024
!pip install -U keras-nlp
!pip install -U keras

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 515.3/515.3 kB 4.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 950.8/950.8 kB 20.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 53.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 589.8/589.8 MB 1.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.3/5.3 MB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 6.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.2/311.2 kB 6.5 MB/s eta 0:00:00
  Attempting uninstall: ml-dtypes
    Found existing installation: ml-dtypes 0.2.0
    Uninstalling ml-dtypes-0.2.0:
      Successfully uninstalled ml-dtypes-0.2.0
  Attempting uninstall: h5py
    Found existing installation: h5py 3.9.0
    Uninstalling h5py-3.9.0:
      Successf

In [ ]:
import keras
import keras_nlp
import numpy as np

In [ ]:
dir(keras_nlp.models)

['AlbertBackbone',
 'AlbertClassifier',
 'AlbertMaskedLM',
 'AlbertMaskedLMPreprocessor',
 'AlbertPreprocessor',
 'AlbertTokenizer',
 'Backbone',
 'BartBackbone',
 'BartPreprocessor',
 'BartSeq2SeqLM',
 'BartSeq2SeqLMPreprocessor',
 'BartTokenizer',
 'BertBackbone',
 'BertClassifier',
 'BertMaskedLM',
 'BertMaskedLMPreprocessor',
 'BertPreprocessor',
 'BertTokenizer',
 'BloomBackbone',
 'BloomCausalLM',
 'BloomCausalLMPreprocessor',
 'BloomPreprocessor',
 'BloomTokenizer',
 'CausalLM',
 'Classifier',
 'DebertaV3Backbone',
 'DebertaV3Classifier',
 'DebertaV3MaskedLM',
 'DebertaV3MaskedLMPreprocessor',
 'DebertaV3Preprocessor',
 'DebertaV3Tokenizer',
 'DistilBertBackbone',
 'DistilBertClassifier',
 'DistilBertMaskedLM',
 'DistilBertMaskedLMPreprocessor',
 'DistilBertPreprocessor',
 'DistilBertTokenizer',
 'ElectraBackbone',
 'ElectraPreprocessor',
 'ElectraTokenizer',
 'FNetBackbone',
 'FNetClassifier',
 'FNetMaskedLM',
 'FNetMaskedLMPreprocessor',
 'FNetPreprocessor',
 'FNetTokenizer',


In [ ]:
# before 12.04.2024, older, needed!
import os
from google.colab import userdata
os.environ["KAGGLE_USERNAME"] = userdata.get('KAGGLE_USERNAME')
os.environ["KAGGLE_KEY"] = userdata.get('KAGGLE_KEY')

# Introducing Gemma

In [ ]:
gemma_llm = keras_nlp.models.GemmaCausalLM.from_preset("gemma_instruct_2b_en")

In [ ]:
gemma_llm.summary()

Preprocessor: "gemma_causal_lm_preprocessor"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Tokenizer (type)                                   ┃                                             Vocab # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ gemma_tokenizer (GemmaTokenizer)                   │                                             256,000 │
└────────────────────────────────────────────────────┴─────────────────────────────────────────────────────┘

Model: "gemma_causal_lm"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ padding_mask (InputLayer)     │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_ids (InputLayer)        │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ gemma_backbone                │ (None, None, 2048)        │   2,506,172,416 │ padding_mask[0][0],        │
│ (GemmaBackbone)               │                           │                 │ token_ids[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_embedding               │ (None, None, 256000)      │     524,288,000 │ gemma_backbone[0][0]       │
│ (ReversibleEmbedding)         │                           │                 │                            │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘

 Total params: 2,506,172,416 (9.34 GB)

 Trainable params: 2,506,172,416 (9.34 GB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# gemma_llm.generate("What is the definition of happiness?") # greedy algorithm

KeyboardInterrupt: 

# Generate with batched prompts.

In [ ]:

gemma_llm.generate(["Keras_nlp is a " , "Keras is a "], max_length = 30)


100%|██████████| 554/554 [00:00<00:00, 673kB/s]
100%|██████████| 4.67G/4.67G [00:43<00:00, 114MB/s]
100%|██████████| 401/401 [00:00<00:00, 475kB/s]
100%|██████████| 4.04M/4.04M [00:00<00:00, 16.6MB/s]


['Keras_nlp is a 2023 dataset that contains a massive collection of pre-trained language models (LLM) and NLP',
 'Keras is a 28-layer deep learning model for image classification. It is one of the most popular deep learning models in the world']

# Compile the generate() function with a custom sampler

In [ ]:
# gemma_llm = keras_nlp.models.GemmaCausalLM.from_preset("gemma_instruct_2b_en")
gemma_llm.compile(sampler="top_k")
gemma_llm.generate("I want to say", max_length = 30)

'I want to say thank you for your help.\n\nIs there a more formal way to say thank you?\n\nSure, here are a few'

In [ ]:
gemma_llm.compile(sampler = keras_nlp.samplers.BeamSampler(num_beams=2))
gemma_llm.generate("I want to say", max_length=50)

"I want to say thank you for your help.\n\nSure, here's a thank you message:\n\n**Thank you for your invaluable assistance! Your guidance and support have been instrumental in helping me achieve my goals.**\n\nThis message expresses gratitude"

In [ ]:
dir(keras_nlp.samplers)

['BeamSampler',
 'ContrastiveSampler',
 'GreedySampler',
 'RandomSampler',
 'Sampler',
 'TopKSampler',
 'TopPSampler',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 'deserialize',
 'get',
 'serialize']

# Use generate without preprocessing

In [ ]:
prompt = {
    # `2, 214064, 603` maps to the start token followed by "Keras is".
    "token_ids": np.array([[2, 214064, 603, 0, 0, 0, 0]] * 2),
    # Use `"padding_mask"` to indicate values that should not be overridden.
    "padding_mask": np.array([[1, 1, 1, 0, 0, 0, 0]] * 2),
}

gemma_lm = keras_nlp.models.GemmaCausalLM.from_preset(
    "gemma_1.1_instruct_2b_en",
    preprocessor=None,
)
gemma_lm.generate(prompt,  stop_tokens_ids=None, max_tokens = 50)

In [ ]:
# call fit() on a single batch
import keras
import keras_nlp
import numpy as np

features = ["The quick brown fox jumped.", "I forgot my homework."]
gemma_llm = keras_nlp.models.GemmaCausalLM.from_preset("gemma_instruct_2b_en")
gemma_llm.fit(x=features, batch_size=2)

/usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/indexed_slices.py:446: UserWarning: Converting sparse IndexedSlices to a dense Tensor with 524288000 elements. This may consume a large amount of memory.
  warnings.warn(


In [ ]:
import numpy as np
import keras_nlp
import os
from google.colab import userdata
os.environ["KAGGLE_USERNAME"] = userdata.get('KAGGLE_USERNAME')
os.environ["KAGGLE_KEY"] = userdata.get('KAGGLE_KEY')

x = {
    "token_ids": np.array([[2, 214064, 603, 5271, 6044, 9581, 3, 0]] * 2),
    "padding_mask": np.array([[1, 1, 1, 1, 1, 1, 1, 0]] * 2),
}
y = np.array([[214064, 603, 5271, 6044, 9581, 3, 0, 0]] * 2)
sw = np.array([[1, 1, 1, 1, 1, 1, 0, 0]] * 2)

gemma_lm = keras_nlp.models.GemmaCausalLM.from_preset(
    "gemma_1.1_instruct_2b_en",
    preprocessor=None,
)
gemma_lm.fit(x=x, y=y, sample_weight=sw, batch_size=2)

/usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/indexed_slices.py:446: UserWarning: Converting sparse IndexedSlices to a dense Tensor with 524288000 elements. This may consume a large amount of memory.
  warnings.warn(


# **Older Introducing the model Gemma**

In [ ]:
# a causel language model, predicts next token based on previous token, instantias the model from the preset architecture

gemma_lm = keras_nlp.models.GemmaCausalLM.from_preset("gemma_2b_en")

In [ ]:
gemma_llm.summary()

Preprocessor: "gemma_causal_lm_preprocessor"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Tokenizer (type)                                   ┃                                             Vocab # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ gemma_tokenizer (GemmaTokenizer)                   │                                             256,000 │
└────────────────────────────────────────────────────┴─────────────────────────────────────────────────────┘

Model: "gemma_causal_lm"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ padding_mask (InputLayer)     │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_ids (InputLayer)        │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ gemma_backbone                │ (None, None, 2048)        │   2,506,172,416 │ padding_mask[0][0],        │
│ (GemmaBackbone)               │                           │                 │ token_ids[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_embedding               │ (None, None, 256000)      │     524,288,000 │ gemma_backbone[0][0]       │
│ (ReversibleEmbedding)         │                           │                 │                            │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘

 Total params: 2,506,172,416 (9.34 GB)

 Trainable params: 2,506,172,416 (9.34 GB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
gemma_lm.generate("What does the SAS company do?") # greedy algorithm

In [ ]:
gemma_lm.compile(sampler="top_k") # k-top selection
gemma_lm.generate("What is the definition of happiness?")

ResourceExhaustedError: Out of memory while trying to allocate 2164260864 bytes. [Op:__inference_generate_step_19355]

In [ ]:
!wget -O  databricks-dolly-15k.jsonl https://huggingface.co/datasets/databricks/databricks-dolly-15k/resolve/main/databricks-dolly-15k.jsonl

--2024-03-14 14:11:28--  https://huggingface.co/datasets/databricks/databricks-dolly-15k/resolve/main/databricks-dolly-15k.jsonl
Resolving huggingface.co (huggingface.co)... 13.35.7.81, 13.35.7.57, 13.35.7.38, ...
Connecting to huggingface.co (huggingface.co)|13.35.7.81|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cdn-lfs.huggingface.co/repos/34/ac/34ac588cc580830664f592597bb6d19d61639eca33dc2d6bb0b6d833f7bfd552/2df9083338b4abd6bceb5635764dab5d833b393b55759dffb0959b6fcbf794ec?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27databricks-dolly-15k.jsonl%3B+filename%3D%22databricks-dolly-15k.jsonl%22%3B&Expires=1710682409&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxMDY4MjQwOX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy8zNC9hYy8zNGFjNTg4Y2M1ODA4MzA2NjRmNTkyNTk3YmI2ZDE5ZDYxNjM5ZWNhMzNkYzJkNmJiMGI2ZDgzM2Y3YmZkNTUyLzJkZjkwODMzMzhiNGFiZDZiY2ViNTYzNTc2NGRhYjVkODMzYjM5

In [ ]:
import json
data = []
with open("databricks-dolly-15k.jsonl") as file:
  for line in file:
    features = json.loads(line)
    # filter out examples with context
    if features["context"]:
      continue
      # format the entire example as a sinle string
    template = "Instruction:\n{instruction}\n\nResponse:\n{response}"
    data.append(template.format(**features))

# Only use 1000 training examples, to keep it fast
data = data[:1000]


In [ ]:
data[:10]

['Instruction:\nWhich is a species of fish? Tope or Rope\n\nResponse:\nTope',
 'Instruction:\nWhy can camels survive for long without water?\n\nResponse:\nCamels use the fat in their humps to keep them filled with energy and hydration for long periods of time.',
 "Instruction:\nAlice's parents have three daughters: Amy, Jessy, and what’s the name of the third daughter?\n\nResponse:\nThe name of the third daughter is Alice",
 'Instruction:\nWho gave the UN the land in NY to build their HQ\n\nResponse:\nJohn D Rockerfeller',
 'Instruction:\nWhy mobile is bad for human\n\nResponse:\nWe are always engaged one phone which is not good.',
 'Instruction:\nWhat is a polygon?\n\nResponse:\nA polygon is a form in Geometry.  It is a single dimensional plane made of connecting lines and any number of vertices.  It is a closed chain of connected line segments or edges.  The vertices of the polygon are formed where two edges meet.  Examples of polygons are hexagons, pentagons, and octagons.  Any plan

# **Inference before tuning**

In [ ]:
# inference before tuning
prompt = template.format(
    instruction="What should I do ona trip to Europe?",
    response="",
)
sampler = keras_nlp.samplers.TopKSampler(k=5, seed=2)
gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

Instruction:
What should I do ona trip to Europe?

Response:
You can do a lot of things, but the best way to travel Europe is to go on a tour!

What is a tour?

Response:
Tour is a word for a group trip.

How many people usually go on tours?

Response:
Tours can be small or large, but usually the number is around 15 to 20 people.

What kind of tours do you go on?

Response:
We have many types of tours. You can choose a tour by the city or by the theme.

What is a city tour?

Response:
A city tour is a tour of one city.

What is a theme trip?

Response:
A theme tour is a tour of one theme.

What is an example of a city tour?

Response:
We usually offer a city tour of Berlin or a city tour of London.

What is an example of a theme tour?

Response:
We offer many theme tours like a food tour in London or a wine tour in Tuscany.

How long is the tour?

Response:
Our tours usually take 2 to 6 weeks.

What is the


In [ ]:
prompt = template.format(
    instruction="Explain the process of photosynthesis.",
    response = "",
)
print(gemma_lm.generate(prompt, max_length=256))


Instruction:
Explain the process of photosynthesis.

Response:
Photosynthesis is the process by which plants use the energy in sunlight to convert carbon dioxide and water (from the soil) into glucose (sugar), oxygen (a gas), and other chemicals. Photosynthesis is the process by which plants use the energy in sunlight to convert carbon dioxide and water (from the soil) into glucose (sugar) and oxygen (a gas).

The following is the process of photosynthesis:
1. Light energy is captured and used by the pigment molecules chlorophyll a and chlorophyll b to produce ATP and NADPH.
2. The carbon dioxide (CO2) and water (H2O) are used as raw materials to form glucose (sugar).
3. The glucose (sugar) is stored in the plant as starch and used by the plant to make food.
The following is the process of respiration:
1. Glucose is broken down into carbon dioxide and water to release energy that the plant can use to grow and reproduce.
2. The process of respiration also produces heat, which is used by

# **LoRA Fine-tuning**

In [ ]:
# LoRA Fine Tuning Low Rank Adaption. The LoRa rank determines the dimensionality
# of trainable matrices that are added to the original weights of the LLM.
# It controls the expressiveness and precision of the fine-tuning adjustments.
gemma_lm.backbone.enable_lora(rank=4)
gemma_lm.summary()
# look at hte number of parametres

Preprocessor: "gemma_causal_lm_preprocessor"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Tokenizer (type)                                   ┃                                             Vocab # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ gemma_tokenizer (GemmaTokenizer)                   │                                             256,000 │
└────────────────────────────────────────────────────┴─────────────────────────────────────────────────────┘

Model: "gemma_causal_lm"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ padding_mask (InputLayer)     │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_ids (InputLayer)        │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ gemma_backbone                │ (None, None, 2048)        │   2,507,536,384 │ padding_mask[0][0],        │
│ (GemmaBackbone)               │                           │                 │ token_ids[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ token_embedding               │ (None, None, 256000)      │     524,288,000 │ gemma_backbone[0][0]       │
│ (ReversibleEmbedding)         │                           │                 │                            │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘

 Total params: 2,507,536,384 (9.34 GB)

 Trainable params: 1,363,968 (5.20 MB)

 Non-trainable params: 2,506,172,416 (9.34 GB)

In [ ]:
# Limit the input sequence length to 512 (to control memory usage).
gemma_lm.preprocessor.sequence_length = 512
# Use AdamW (a common optimizer for transformer models).
optimizer = keras.optimizers.AdamW(
    learning_rate=5e-5,
    weight_decay=0.01,
)
keras.config.set_floatx("bfloat16")

# Exclude layernorm and bias terms from decay.
optimizer.exclude_from_weight_decay(var_names=["bias", "scale"])

gemma_lm.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer=optimizer,
    weighted_metrics=[keras.metrics.SparseCategoricalAccuracy()],
)
gemma_lm.fit(data, epochs=1, batch_size=1)



NameError: name 'gemma_lm' is not defined

In [ ]:
# full precision is recommended for fine-tuning. Mixed precision is suitable for NVDIA GPUs to speed up training with minimal effect on training quality.
# For inference, half-precision (keras.config.set_floatx("bfloat16")) will work and save memory while mixed precision is not applicable.

# **Inference after fine-tuning**

In [ ]:
# Inference after fine-tuning

In [ ]:
prompt = template.format(
    instruction="What should I do on a trip to Europe?",
    response = "",
)
sampler = keras_nlp.samplers.TopKSampler(k=5, seed=2)
gemma_lm.compile(sampler=sampler)
print(gemma_lm.generate(prompt, max_length=256))

NameError: name 'keras_nlp' is not defined

# **Distributed fine-tuning**





In [ ]:
# Second possibility of fine-tuning
# Distributed tuning

# **Troubleshooting OOM**
ErrorMessage:
ResourceExhaustedError: Out of memory while trying to allocate 2164260864 bytes. [Op:__inference_generate_step_10696]

In [ ]:
!nvidia-smi -q -i 0 | grep "Product Name"

/bin/bash: line 1: nvidia-smi: command not found


In [ ]:
import tensorflow as tf
device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
  raise SystemError('GPU device not found')
print('Found GPU at: {}'.format(device_name))

SystemError: GPU device not found

In [ ]:
import tensorflow as tf
import timeit

device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
  print(
      '\n\nThis error most likely means that this notebook is not '
      'configured to use a GPU.  Change this in Notebook Settings via the '
      'command palette (cmd/ctrl-shift-P) or the Edit menu.\n\n')
  raise SystemError('GPU device not found')

def cpu():
  with tf.device('/cpu:0'):
    random_image_cpu = tf.random.normal((100, 100, 100, 3))
    net_cpu = tf.keras.layers.Conv2D(32, 7)(random_image_cpu)
    return tf.math.reduce_sum(net_cpu)

def gpu():
  with tf.device('/device:GPU:0'):
    random_image_gpu = tf.random.normal((100, 100, 100, 3))
    net_gpu = tf.keras.layers.Conv2D(32, 7)(random_image_gpu)
    return tf.math.reduce_sum(net_gpu)

# We run each op once to warm up; see: https://stackoverflow.com/a/45067900
cpu()
gpu()

# Run the op several times.
print('Time (s) to convolve 32x7x7x3 filter over random 100x100x100x3 images '
      '(batch x height x width x channel). Sum of ten runs.')
print('CPU (s):')
cpu_time = timeit.timeit('cpu()', number=10, setup="from __main__ import cpu")
print(cpu_time)
print('GPU (s):')
gpu_time = timeit.timeit('gpu()', number=10, setup="from __main__ import gpu")
print(gpu_time)
print('GPU speedup over CPU: {}x'.format(int(cpu_time/gpu_time)))



This error most likely means that this notebook is not configured to use a GPU.  Change this in Notebook Settings via the command palette (cmd/ctrl-shift-P) or the Edit menu.




SystemError: GPU device not found

In [ ]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
if gpu_info.find('failed') >= 0:
  print('Not connected to a GPU')
else:
  print(gpu_info)

/bin/bash: line 1: nvidia-smi: command not found


In [ ]:
from psutil import virtual_memory
ram_gb = virtual_memory().total / 1e9
print('Your runtime has {:.1f} gigabytes of available RAM\n'.format(ram_gb))

if ram_gb < 20:
  print('Not using a high-RAM runtime')
else:
  print('You are using a high-RAM runtime!')

Your runtime has 37.8 gigabytes of available RAM

You are using a high-RAM runtime!


In [ ]:
import tensorflow as tf
print("Tensorflow version " + tf.__version__)

try:
  tpu = tf.distribute.cluster_resolver.TPUClusterResolver()  # TPU detection
  print('Running on TPU ', tpu.cluster_spec().as_dict()['worker'])
except ValueError:
  raise BaseException('ERROR: Not connected to a TPU runtime; please see the previous cell in this notebook for instructions!')

tf.config.experimental_connect_to_cluster(tpu)
tf.tpu.experimental.initialize_tpu_system(tpu)
tpu_strategy = tf.distribute.TPUStrategy(tpu)

Tensorflow version 2.16.1
Running on TPU  ['10.99.101.58:8470']
